wgpu-core/src/lock/mod.rs

```diff
@@ -0,0 +1,40 @@
+//! Instrumented lock types.
+//!
+//! This module defines a set of instrumented wrappers for the lock
+//! types used in `wgpu-core` ([`Mutex`], [`RwLock`], and
+//! [`SnatchLock`]) that help us understand and validate `wgpu-core`
+//! synchronization.
+//!
+//! - The [`ranked`] module defines lock types that perform run-time
+//!   checks to ensure that each thread acquires locks only in a
+//!   specific order, to prevent deadlocks.
+//!
+//! - The [`vanilla`] module defines lock types that are
+//!   uninstrumented, no-overhead wrappers around the standard lock
+//!   types.
+//!
+//! (We plan to add more wrappers in the future.)
+//!
+//! If the `validate-locks` feature is enabled, `wgpu-core` uses the
+//! [`ranked`] module's locks. We hope to make this the default for
+//! debug builds soon.
+//!
+//! Otherwise, `wgpu-core` uses the [`vanilla`] module's locks.
+//!
+//! [`Mutex`]: parking_lot::Mutex
+//! [`RwLock`]: parking_lot::RwLock
+//! [`SnatchLock`]: crate::snatch::SnatchLock
+
+pub mod rank;
```

Comment:
nitpick: We're still unconditionally using `allow(dead_code)` in the `rank` module. I think we should have a similar `#[cfg_attr(feature = "validate-locks", allow(dead_code))]` here instead?

Reply:
Hmm, it's funny, it's only the fields of `LockRank` that aren't read. We certainly use the rank constants unconditionally in `wgpu-core`.

Reply:
It turns out that... the `cfg_attr` conditions were swapped??? Anyway, now that they're fixed, there are no extraneous `allow`s in `wgpu-core/src/lock`.